### ISSUE TITLE
Code questions

### ISSUE THREAD

**Issue by pILLOW-1** [Never-wx/GCD#3]
Hi, great work! I have several code questions listed below.
1. scripts/select_categories_nstep_train.py

![Image](https://github.com/user-attachments/assets/6f027256-27e4-4508-81cb-364baa271e9e)
    
If you want to divide instances_train2017_40-79.json into 4 steps, then there may be a problem. 
The num_cls is now 80 (refer to the implementation of select_categories_2step.py), so the cls_per_phase will be 20. But what we want is 10 actually. Additionally, in line 28-29, the "start" will begin from 0 rather than 40. 
This question is closely related to the following questions.

2. configs/gdino_inc/40+10_4/gdino_inc_40+10_4_40-49_gcd_scratch_coco.py
    2.1 I don't understand "ori_config_file" in line 9. 
    2.2 Why the ann_file is 'annotations/70+10/instances_train2017_70-79.json' in line 45?
3. mmdet/datasets/coco_increment.py
   Here comes my core question. Let us say the setting is 40+10_4 and the current training stage is 40-49. In line 27-28, self.metainfo['classes'] is 0-79 or 0-49 or 40-49 ? According to line 6, the start is 40, so self.metainfo['classes'] could be 0-79 or 0-49 so that the length is larger than 40. Also, self._metainfo['new_classes'] is obtained from self.metainfo['classes'][self.start:], which means self.metainfo['classes'] could only be 0-49. But according to the implementation of select_categories_nstep_train.py, the length of 'categories' written to the json is less than 50.

I am really confused about the implementation mentioned above and hope the author can provide a clear explanation for the questions. Take the setting of 40+10_4 for example, I want to figure out: 1. What should the "categories" in the json in each phase be like? 2. What the self.metainfo['classes'] in the dataset in each phase should be like?

Thanks a lot!

**Comment by pILLOW-1** [comment]
Maybe change select_categories_nstep_train.py
line 21 to: num_cls = len(dataset['categories']) - Start
line 28-30 to:
start = Start + phase * cls_per_phase
end = Start + (phase + 1) * cls_per_phase
sel_cats = dataset['categories'][:end] # ?

where "Start" means the end of the last phase and is 40 under the setting of 40+10_4

After changing like this, I think the other parts (e.g., config and coco_increment) make sense

**Comment by pILLOW-1** [comment]
Maybe change select_categories_nstep_train.py
line 21 to: num_cls = len(dataset['categories']) - Start
line 28-29 to:
start = Start + phase * cls_per_phase
end = Start + (phase + 1) * cls_per_phase
line 43 to: sel_dataset = {'categories': dataset['categories'][:end], 'annotations': sel_anno, 'images': sel_images}

where "Start" means the end of the last phase and is 40 under the setting of 40+10_4

After changing like this, I think the other parts (e.g., config and coco_increment) make sense

**Comment by pILLOW-1** [comment]
Maybe change select_categories_nstep_train.py
line 21 to: num_cls = len(dataset['categories']) - Start
line 28-29 to:
start = Start + phase * cls_per_phase
end = Start + (phase + 1) * cls_per_phase
line 43 to: sel_dataset = {'categories': dataset['categories'][:end], 'annotations': sel_anno, 'images': sel_images}

where "Start" means the end of the last phase and is 40 under the setting of 40+10_4

change select_categories_nstep_val.py
line 35 to: sel_dataset = {'categories': sel_cats, 'annotations': sel_anno, 'images': sel_images} 

After changing like this, I think the other parts (e.g., config and coco_increment) make sense

**Comment by pILLOW-1** [comment]
Another question is the misalignment between code and implementation details in the paper:

![Image](https://github.com/user-attachments/assets/0866fa77-1970-43fa-a103-fc1232e5cd2f)

![Image](https://github.com/user-attachments/assets/7d063f98-e12e-436c-851f-801e2b381c92)

![Image](https://github.com/user-attachments/assets/808cadb6-c044-461b-9553-e024c158898a)

1. The paper says the language model is frozen after the first phase, but it is frozen in the first phase as well.
2. The learning rate is inconsistent.

**Comment by Never-wx** [comment]
Thank you for pointing out the bug.
Q1: For the training annotation file instances_train2017_40-79.json, it only includes categories 40–79 (excluding previous ones), so num_cls = len(dataset['categories']) = 40 is correct. However, there is indeed a bug in the output file naming — the files should be named 40-49, 50-59, etc., based on the actual category ranges.

Q2: The ori_config_file refers to the teacher model’s original config, which should not include the distillation part. As for the annotation file, it should correspond to 40-49. I’ll double-check that to confirm.

Q3: In coco_increment.py (lines 86–93), cur_text means only the current text (i.e., 40–49) is provided. If we are fine-tuning without distillation, cur_text mode is used. If distillation is involved, we use the full text (e.g., 0–49), and map annotations to global categories. In lines 31–32, we define self._metainfo['ori_classes'] = [0:39] and self._metainfo['new_classes'] = [40:49].
The instances_train2017_40-49.json contains only categories 40–49, while instances_val2017_0-49.json contains 0–49. The input text setting is controlled by the config (cfg), not directly by the annotation file — this logic is implemented in lines 153–161 of coco_increment.py.

![Image](https://github.com/user-attachments/assets/e3c869e9-686c-4f5a-87ef-6632a7f17e45)

Q4:   The language_model_frozen parameter should be set to False. The learning rate for the base phase is set to 1e-4, and 1e-5 for the vision/language backbone. For the subsequent incremental phases, the learning rate settings remain consistent with those described in our paper.   

The base cfg will be updated. Besides, the base weights and splits are provided in this repo.

**Comment by pILLOW-1** [comment]
Many thanks! But I still have some questions based on your reply.
Q1: I know that num_cls in select_categories_nstep_train.py should be 40, the question is len(dataset['categories']) != 40. In select_categories_2step.py, the categories written to instances_train2017_40-79.json is actually 80. Please double-check.
![Image](https://github.com/user-attachments/assets/b3cf3b24-9808-4450-b843-4664e2f9456b)

Q2: Why the teacher is the same for 40-49, 50-59, 60-69 and 70-79? For example, shouldn't the teacher for 50-59 be 0-49 rather than 0-39?

Q3: In coco_increment.py, maybe self._metainfo['new_classes'] = self.metainfo['classes'][self.start:self.end] ?
![Image](https://github.com/user-attachments/assets/982ec901-32ee-4579-b548-1bff0adfc8d7)

Q4: In configs/gdino_inc/40+10_4/gdino_inc_40+10_4_50-59_gcd_scratch_coco.py, maybe a typo for load_from? Where is "adaptive"? Same question for 60-69 and 70-79.
![Image](https://github.com/user-attachments/assets/2590fd1d-da2f-4246-94e3-6c0a16244cae)

Q5: I think there should be some modifications to select_categories_2step.py and select_categories_nstep_val.py to make the whole discussion reasonable. Let me explain. 
First, please refer to line 55 in the figure in Q1. I think "sel_cats" rather than dataset['categories'] should be written to "categories" so that  instances_train2017_40-79.json can only include categories 40–79. Also, to make the code look more robust, it's better to add an 'if' condition to exclude the val_output in the second step since there is no need to do that (the val part in 40-79 and 70-79 config files already indicates that). 
![Image](https://github.com/user-attachments/assets/2224f20f-4d15-405b-8a70-e7891099da41)
Second, please refer to line 35 in scripts/select_categories_nstep_val.py. It is "sel_cats" rather than dataset['categories'] that should be written to "categories" so that  instances_val2017_0-49.json can include categories 0–49 rather than 0-79 (same for 0-59 and 0-69).
After changing like this, I think the updated select_categories_nstep_train.py and the subsequent config and coco_increment all make sense.

Q6: Where can I check the "num_classes" of the detector? I mean I didn't find "num_classes" in config file.

Looking forward to your answer.

**Comment by pILLOW-1** [comment]
Another question is that why is the language model fintuned in the first phase? Can I freeze it throughout all phases?

**Comment by Never-wx** [comment]
Q1Q5:  In select_categories_2step.py (line 55), the line should be updated to sel_dataset = {'categories': sel_cats, 'annotations': sel_anno, 'images': sel_images}. The same change should be applied to select_categories_nstep_val.py.


I’ve tested the result using instances_val2017_0-49.json (which includes only categories 0–49), and the output is correct. The previous implementation mistakenly included all 80 categories in the JSON file, even though annotations for categories 50–79 were not present. This led to the same performance results, which is why the bug was initially overlooked.

**Performance with 0–49 categories** 
![Image](https://github.com/user-attachments/assets/1d545ed9-42e1-4ded-8cbf-efb138f30c47)
**Performance with 0–79 categories included in the JSON**
![Image](https://github.com/user-attachments/assets/2e313f19-2532-40fe-a57f-7e7426bad8c0)

Q2:  The teacher model runs in inference mode, so the config doesn’t need to be changed (it only defines the model architecture). At the beginning of training, the teacher loads the same weights as the student model. This is implemented in mmdet/engine/hooks/inc_distn_hook.py. The ori_text is not defined in the config file—it is determined in coco_increment.py, lines 153–161.

![Image](https://github.com/user-attachments/assets/d9efa7ec-06fa-439c-b2f2-f5ddd63a2d10)

Q3:   The model makes predictions based on input text.
The attributes self._metainfo['new_classes'] and self._metainfo['old_classes'] are only used for evaluation purposes—to compute AP for old and new categories. This logic is handled in mmdet/evaluation/metrics/coco_inc_metric.py.

![Image](https://github.com/user-attachments/assets/fba9471f-13a4-4771-ac8d-24c1f31cfe5f)

![Image](https://github.com/user-attachments/assets/05e2a40e-e9c0-42cc-8655-dac17d202bc1)

Q4:  I have modified the config, It should be last step checkpoint namely ./work_dirs/gdino_inc_40+10_4_40-49_gcd_scratch_coco/epoch_12.pth.

Q6:   num_classes is replaced by trunc_class = [start:end] in mmdet/models/dense_heads/gdino_head_inc.py

![Image](https://github.com/user-attachments/assets/d91a8cde-f9c6-4922-ad5c-38b81ff48406)

Q7:  For the base phase, we follow the standard MMDetection training pipeline. Updating the language backbone during this phase leads to better AP.  However, during the incremental phase, we found that freezing the language backbone helps maintain a more stable feature space.

**Comment by pILLOW-1** [comment]
Many thanks! You solved all my questions! @Never-wx 👍

**Comment by pILLOW-1** [comment]
One more question: If I use another detection model (e.g., DETR), can I directly use the current annotation files and config files ( changing model config and removing your method logic)?
